nifi-extension-bundles/nifi-atlassian-bundle/nifi-atlassian-extensions/src/main/java/org/apache/nifi/atlassian/bitbucket/BitbucketRepositoryClient.java

```diff
@@ -348,39 +348,53 @@ private Iterator<JsonNode> getFiles(final String branch, final String resolvedPa
+        final String errorMessage = String.format("Error while listing content for repository [%s] on branch %s at path %s", repoName, branch, resolvedPath);
 
-        if (response.statusCode() != HttpURLConnection.HTTP_OK) {
-            throw new FlowRegistryException(
-                    String.format("Error while listing content for repository [%s] on branch %s at path %s: %s", repoName, branch, resolvedPath, getErrorMessage(response)));
-        }
-
-        final JsonNode jsonResponse;
-        try {
-            jsonResponse = this.objectMapper.readTree(response.body());
-        } catch (IOException e) {
-            throw new FlowRegistryException("Could not parse response from Bitbucket API", e);
-        }
-        return jsonResponse.get("values").elements();
+        return paginateAndCollect(uri.toString(), errorMessage);
     }
 
     private Iterator<JsonNode> getListCommits(final String branch, final String path) throws FlowRegistryException {
         // retrieve latest commit for that branch
         // https://api.bitbucket.org/2.0/repositories/{workspace}/{repoName}/commits/{branch}
         final URI uri = getUriBuilder().addPathSegment("commits").addPathSegment(branch).addQueryParameter("path", path).build();
-        final HttpResponseEntity response = this.webClient.getWebClientService().get().uri(uri).header(AUTHORIZATION_HEADER, authToken.getAuthzHeaderValue()).retrieve();
+        final String errorMessage = String.format("Error while listing commits for repository [%s] on branch %s", repoName, branch);
 
-        if (response.statusCode() != HttpURLConnection.HTTP_OK) {
-            throw new FlowRegistryException(String.format("Error while listing commits for repository [%s] on branch %s: %s", repoName, branch, getErrorMessage(response)));
-        }
+        return paginateAndCollect(uri.toString(), errorMessage);
+    }
 
-        final JsonNode jsonResponse;
-        try {
-            jsonResponse = this.objectMapper.readTree(response.body());
-        } catch (IOException e) {
-            throw new FlowRegistryException("Could not parse response from Bitbucket API", e);
+    private Iterator<JsonNode> paginateAndCollect(final String url, final String errorMessage) throws FlowRegistryException {
+        final List<JsonNode> allValues = new ArrayList<>();
+        String newUrl = url;
+        while (newUrl != null) {
+            HttpResponseEntity response = webClient.getWebClientService()
+                    .get()
+                    .uri(URI.create(newUrl))
+                    .header(AUTHORIZATION_HEADER, authToken.getAuthzHeaderValue())
+                    .retrieve();
+
+            if (response.statusCode() != HttpURLConnection.HTTP_OK) {
+                throw new FlowRegistryException(String.format(errorMessage + ": %s", getErrorMessage(response)));
```

Comment:
This combination of string formatting and concatenation is a bit hard to read. Since the method is private, the input `errorMessage` could be passed as `errorMessageFormat` with a `%s` placeholder, and then this could be simplified to `errorMessageFormat.formatted(getErrorMessage(response))`. Although I also recommend declaring `final String responseErrorMessage = getErrorMessage(response)` for readability.